# COGS 108 - Data Checkpoint

## Names

- Ava Hamedi
- Marc Mendoza
- Jonathan Park
- Daniel Renteria
- Siena Rivera

<a id='research_question'></a>
# Research Question

Did California Air Quality significantly improve in different areas due to the COVID-19 Pandemic and the change in car traffic volume? 

# Dataset(s)

### Main Datasets

- **Annual Air Quality by County**
  - **Name:** annual_aqi_by_county_YEAR.csv
  - [Link to the Dataset](https://aqs.epa.gov/aqsweb/airdata/download_files.html#Annual)
  - **Description:** This dataset contains data and information on Annual Summary of Air Quality through years before 2000 until today. All Data use AQI (Air Quality Index) when applicable. The Measurement AQI is calculated based on the average concentration of a particular pollutant measured over a standard time interval (24 hours for most pollutants, 8 hours for carbon monoxide and ozone). We will be using this dataset to help understand and show the trend of Air Quality throughout years 2007 until today. This way, we can get a trend of how AQI has changed over the years before and through COVID-19.
  - **Source:** EPA: AirData<br>
<br>
- **Annual Average Daily Traffic (AADT) Volumes (1)**
  - **Name:** traffic_volumes_YEAR.csv
  - [Link to the Dataset](https://dot.ca.gov/programs/traffic-operations/census)
  - **Description:** In addition to Air Quality, our team wanted to look into the relationship between Air Quality and Car and Vehicular Traffic. This dataset contains data and information on Annual Average Daily Traffic Volumes through years 2017 until 2022. This and the following Datasets are the same kind of data but with different years. We will be using this dataset to help understand the average amount of Annual daily traffic through the California area. Although this dataset is more narrow, this will help us have a more narrow scope of information. In comparison to the one below which is the former years. In addition to gathering and cleaning this dataset, we will also be computing the average (mean) of the (Back and Forward Traffic) columns.
  - **Source:** AQICN.org<br>
  <br>
- **Annual Average Daily Traffic (AADT) Volumes (2)**
  - **Name:** traffic_volumes_YEAR.csv
  - [Link to the Dataset](https://data.ca.gov/dataset/fb62fd37-38e5-40a5-89d1-cb58ae12f244/resource/0f021a84-ded0-488c-93b5-f331759ad9fd/download/aadt_2007-2017_shapefiles.gdb.zip)
  - **Description:** This dataset contains data and information on Annual Average Daily Traffic Volumes through years 2007 until 2017. This and the former Datasets are the same kind of data but with different years. We will be using this dataset to help understand the average amount of Annual daily traffic through the California area. This dataset is more diverse, with over 10 years of data and information for our group to use, this will help us have a more wide scope of information. In addition to gathering and cleaning this  dataset, we will also be computing the average (mean) of the (Back and Forward Traffic) columns.
  - **Source:** AQICN.org<br>
<br>
- **Annual Gasoline Sales by City**
  - **Name:** gas_sales_by_city.csv
  - [Link to the Dataset](https://www.energy.ca.gov/media/5869)
  - **Description:** The chosen dataset contains data and information on Annual Gasoline Sales Volumes per City throughout the years 2010 until 2020. We will be cleaning the data and only provide data through years 2011 until 2020 to fit our research question time frame. We will be using this dataset to help understand the usage of gasoline throughout California, and use this to find a correlation to the Air Quality throughout California. Although the data is relatively clean, there are some changes we would like to make to make it easier to understand and work with. Since this dataset has the perfect time frame relative to what we are researching, it will provide our group the 10 years of data necessary to find meaningful relationships. Through this, we can get an idea of how Gasoline Sales Data has changed over the years prior to and during the COVID-19 pandemic, and how this has ultimately affected the AQI throughout California.
  -**Source:** CEC: energy.ca.gov <br>

<br>
 

**Combining Data**: The datasets we have chosen will be combined and subsequently analyzed to give a measure of the overall relationship between Traffic, Air Quality, and Gasoline usage. From there, we want to look further into the addition of the COVID-19 Pandemic. Our team will be analyzing the causal relationship between the effects of COVID-19 on Air Quality stemming from a change in Car and Vehicular Traffic as well as a change in Gasoline sales. All of our information will be combined into one, larger dataset for the California Area. To answer our Research Question, our main priority is getting all of our data cleaned and ready for analysis.

# Setup

In [92]:
# First make sure to have all packages installed on device (pip install pandas, numpy, seaborn, matplotlib, and etc)
# import all packages after installing pip
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt


# Our Warnings when importing code will make it harder to concentrate on what is important
import warnings
warnings.filterwarnings('ignore')

# Data Cleaning

## Table of Contents

- Air Quality in various regions
- Car Traffic 
- Gasoline Consumption 


## Air Quality (AQI) Data

### I. Annual Air Quality (AQI) by County
First we will focus on the Air Quality (AQI) Data given to the team by the EPA. These dataset contained information for the years of 2010-2021. We only want to focus on years 2010-2021, so we will be removing data that does not fit into our timeline. All Data use AQI (Air Quality Index). The Measurement AQI is calculated based on the average concentration of a particular pollutant measured over a standard time interval (24 hours for most pollutants, 8 hours for carbon monoxide and ozone). We will be using this dataset to help understand and show the trend of Air Quality throughout years 2010 until today. This way, we can get a trend of how AQI has changed over the years before and through COVID-19.


In [111]:
#importing data
aqi_county_path = ['datasets/annual_aqi_by_county_2010.csv', 
                 'datasets/annual_aqi_by_county_2011.csv', 
                 'datasets/annual_aqi_by_county_2012.csv', 
                 'datasets/annual_aqi_by_county_2013.csv', 
                 'datasets/annual_aqi_by_county_2014.csv', 
                 'datasets/annual_aqi_by_county_2015.csv', 
                 'datasets/annual_aqi_by_county_2016.csv', 
                 'datasets/annual_aqi_by_county_2017.csv', 
                 'datasets/annual_aqi_by_county_2018.csv',
                 'datasets/annual_aqi_by_county_2019.csv', 
                 'datasets/annual_aqi_by_county_2020.csv', 
                 'datasets/annual_aqi_by_county_2021.csv']

In [112]:
#parsing data
aqi_county_frames = []
for path in aqi_county_path:
    aqi_county_frames.append(pd.read_csv(path))
aqi_county = pd.concat(aqi_county_frames, ignore_index = True)

In [113]:
#columns to remove from files
aqi_county_bad = ['Days with AQI', 'Good Days', 'Moderate Days', 'Unhealthy for Sensitive Groups Days',
                  'Unhealthy Days', 'Very Unhealthy Days', 'Hazardous Days', 'Days CO', 'Days NO2',
                  'Days Ozone', 'Days SO2', 'Days PM2.5', 'Days PM10']

In [114]:
#dropping columns
aqi_county.drop(aqi_county_bad, axis = 1, inplace = True)

In [115]:
#dropping rows
aqi_county_CA = aqi_county[aqi_county['State']=='California']

In [116]:
#resetting index
aqi_county_CA.reset_index(drop=True, inplace = True)

In [117]:
aqi_county_CA.head()

State     County  Year  Max AQI  90th Percentile AQI  Median AQI
0  California    Alameda  2010      179                   68          43
1  California     Amador  2010      151                   64          35
2  California      Butte  2010      126                   84          47
3  California  Calaveras  2010      154                   84          41
4  California     Colusa  2010      119                   49          38

#### Columns
- State: The state in which the Air Quality was measured. In our case, it is only California.
- County: The county (of California) in which the Air Quality was measured.
- Year: The year the Air Quality was measured.
- Max AQI: The maximum AQI (worst air quality) that year.
- 90th Percentile AQI: The 90th percentile AQI value that year.
- Median AQI: The median AQI that year.

## Traffic Data

### II. Annual Daily Traffic Data 
In addition to Air Quality, our team wanted to look into the relationship between Air Quality and Car and Vehicular Traffic. The chosen datasets contain data and information on Annual Average Daily Traffic Volumes through years 2007 until 2020. We will be using this dataset to help understand the average amount of Annual daily traffic through the state of California. We found this information via data.ca.gov, through the data that CalTrans gathers on traffic volumes. We only want to focus on a more narrow area and scope, so we will be cleaning up the datasets to help provide us with the best information. This way, we can also get a trend of how Annual Daily Traffic Data has changed over the years before and through COVID-19. We will also be showcasing the data in ways that are most intuitive. We will eventually be  grouping and taking the average (mean) of the (Back and Forward Traffic) columns. We will also only look at the AADT, the annual average, so we will drop all columns pertaining to daily measured peaks or monthly measures.


In [120]:
#importing data               
traffic_path_1 = ['datasets/traffic_volumes_2007.csv',
               'datasets/traffic_volumes_2008.csv',
               'datasets/traffic_volumes_2009.csv',
               'datasets/traffic_volumes_2010.csv',
               'datasets/traffic_volumes_2011.csv',
               'datasets/traffic_volumes_2012.csv',
               'datasets/traffic_volumes_2013.csv',
               'datasets/traffic_volumes_2014.csv',
               'datasets/traffic_volumes_2015.csv',
               'datasets/traffic_volumes_2016.csv',
               'datasets/traffic_volumes_2017.csv']
traffic_path_2 = ['datasets/traffic_volumes_2018.csv',
               'datasets/traffic_volumes_2019.csv']
traffic_path_3 = 'datasets/traffic_volumes_2020.csv'

In [121]:
#parsing data
traffic_frames_1 = []
for path in traffic_path_1:
    traffic_frames_1.append(pd.read_csv(path))
traffic_1 = pd.concat(traffic_frames_1, ignore_index= True)

In [122]:
traffic_frames_2 = []
for path in traffic_path_2:
    traffic_frames_2.append(pd.read_csv(path))
traffic_2 = pd.concat(traffic_frames_2, ignore_index= True)

In [123]:
traffic_3 = pd.read_csv(traffic_path_3)

In [124]:
#columns to keep from files
traffic_keep_1 = ['County', 'Back_AADT', 'Ahead_AADT']
traffic_keep_2 = ['CNTY', 'BACK_AADT', 'AHEAD_AADT']
traffic_keep_3 = ['COUNTY', 'BACK_AADT', 'AHEAD_AADT']

In [125]:
#dropping, renaming columns
traffic_1_cond = traffic_1[traffic_keep_1]
traffic_2_cond = traffic_2[traffic_keep_2].rename(columns={'CNTY':'County', 'BACK_AADT':'Back_AADT', 'AHEAD_AADT':'Ahead_AADT'})
traffic_3_cond = traffic_3[traffic_keep_3].rename(columns={'COUNTY':'County', 'BACK_AADT':'Back_AADT', 'AHEAD_AADT':'Ahead_AADT'})

In [126]:
#combining dataframes
traffic_frames_final = [traffic_1_cond, traffic_2_cond, traffic_3_cond]
traffic = pd.concat(traffic_frames_final, ignore_index= True)

In [127]:
#dropping null rows
traffic.dropna(axis=0, inplace= True)

In [128]:
#renaming row values
full_name = {'ALA':'Alameda','ALP':'Alpine','AMA':'Amador','BUT':'Butte', 'CAL':'Calaveras', 'CC':'Contra Costa','COL':'Colusa','DN':'Del Norte','ED':'El Dorado','FRE':'Fresno','GLE':'Glenn','HUM':'Humboldt',
            'IMP':'Imperial','INY':'Inyo','KER':'Kern','KIN':'Kings','LA':'Los Angeles','LAK':'Lake','LAS':'Lassen','MAD':'Madera','MEN':'Mendocino','MER':'Merced','MNO':'Mono','MOD':'Modoc','MON':'Monterey',
            'MPA':'Mariposa','MRN':'Marin','NAP':'Napa','NEV':'Nevada','ORA':'Orange','PLA':'Placer','PLU':'Plumas','RIV':'Riverside','SAC':'Sacramento','SB':'Santa Barbara','SBD':'San Bernardino',
            'SBT':'San Benito','SCL':'Santa Clara','SCR':'Santa Cruz','SD':'San Diego', 'SF': 'San Francisco','SHA':'Shasta','SIE':'Sierra','SIS':'Siskiyou', 'SJ':'San Joaquin', 'SLO': 'San Luis Obispo',
            'SM':'San Mateo','SOL':'Solano','SON':'Sonoma','STA':'Stanislaus','SUT':'Sutter','TEH':'Tehama','TRI':'Trinity','TUL':'Tulare','TUO':'Tuolumne','VEN':'Ventura','YOL':'Yolo','YUB':'Yuba'}
traffic['County'] = traffic['County'].replace(full_name)

In [138]:
#grouping by county, then adding mean AADT column
traffic_counties = traffic.groupby('County').sum().reset_index()[1:].reset_index().drop(columns='index')
AADT_mean = (traffic_counties['Back_AADT'] + traffic_counties['Ahead_AADT'])/2
traffic_counties['Mean_AADT'] = AADT_mean
traffic_counties.head()

County    Back_AADT   Ahead_AADT    Mean_AADT
0    Alameda  301232550.0  294978200.0  298105375.0
1     Alpine     342560.0     343275.0     342917.5
2     Amador    4694750.0    4292950.0    4493850.0
3      Butte   21355250.0   21016740.0   21185995.0
4  Calaveras    3699350.0    3554910.0    3627130.0

### Columns
- County: California county where the traffic volume was measured.
- Back_AADT: The total count of vehicles measured at the "back" of a measured road (before traffic light or intersection).
- Ahead_AADT: The total count of vehicles measured at the "front" of a measured road (past traffic light or intersection).
- Mean_AADT: The mean of Back_AADT and Ahead_AADT.

## Gasoline Sales Data

### III. Annual Gasoline Sales Data 
In addition to Vehicle Traffic, our team wanted to look into the relationship between Air Quality and Gasoline usage. The chosen dataset contains data and information on Annual Gasoline Sales Volumes per City throughout the years 2010 until 2020, however, we will be cleaning the data and only provide data through years 2011 until 2020. We will be using this dataset to help understand the usage of gasoline throughout California, and use this to find a correlation to the Air Quality. Although the data is relatively clean, there are some changes we would like to make to make it easier to understand and work with. Through this, we can get an idea of how Gasoline Sales Data has changed over the years prior to and during the COVID-19 pandemic, and how this has ultimately affected the AQI throughout California.


In [139]:
# Import Data
gas_sales = pd.read_csv('datasets/gas_sales_by_city.csv')

# Drop 2010 Column
gas_sales.drop(labels="2010", axis = 1, inplace = True)

#Drop null values (if any)
gas_sales.dropna(axis=0, inplace= True)

# Format City Column
gas_sales['City'] = gas_sales['City'].str.title()
gas_sales


City            2011            2012            2013  \
0    Los Angeles    576,975,774     569,572,568     513,096,586    
1      San Diego    435,507,017     418,994,094     350,573,480    
2     Sacramento    259,938,992     260,349,538     255,080,126    
3       San Jose    299,816,301     292,742,725     278,676,054    
4    Bakersfield    193,751,205     193,930,997     178,563,239    
..           ...             ...             ...             ...   
367     Paradise      9,459,933       6,628,310       6,579,072    
368    Dos Palos      1,963,102       2,158,303       3,144,116    
369       Quincy      4,531,877       2,418,894       1,577,234    
370  Angels Camp      2,797,978       1,663,664       1,045,570    
371      Needles      2,450,049       1,585,251       1,794,034    

               2014            2015            2016            2017  \
0      462,203,720     489,927,305     577,624,448     597,482,538    
1      382,951,006     429,738,409     458,149,963     478,402,573    
2      174,054,379     255,968,575     278,560,889     283,805,191    
3      218,224,534     286,512,916     324,059,453     318,609,135    
4      168,800,299     181,022,100     221,473,122     216,668,139    
..              ...             ...             ...             ...   
367      6,023,551       7,181,204      10,214,292      10,009,953    
368      3,119,857       2,549,044       3,678,522       3,361,169    
369      2,659,001       2,221,433       1,927,929       2,672,375    
370      1,470,884       1,913,018       2,215,584       2,286,907    
371      1,828,995       2,259,731       2,065,422       2,129,009    

               2018            2019            2020  
0      553,325,286     551,182,317     434,124,374   
1      463,772,772     454,951,806     351,446,047   
2      264,019,029     278,151,293     245,128,423   
3      298,305,348     311,325,761     234,628,392   
4      204,619,282     204,009,437     190,635,868   
..              ...             ...             ...  
367      7,693,650       1,599,028       2,930,659   
368      2,728,566       2,379,250       2,157,239   
369      3,464,343       1,831,817       2,112,113   
370      2,178,426       2,288,752       1,774,643   
371      2,201,664       2,170,556       1,564,649   

[372 rows x 11 columns]

The units being used to track gasoline sales over the 10 years in the dataset are in 1000 gallon units. There are a total of 11 columns where one is for the City names, and the other ten are for the years 2011 through 2020, the years where gasoline sales were being tracked. There are a total of 372 rows for the cities that actually had gasoline sales data, with their respective gasoline sales per year.